Winleafs.Wpf/Views/Layout/CreateEffectWindow.xaml.cs

```diff
@@ -0,0 +1,262 @@
+﻿using Newtonsoft.Json;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using System.Windows;
+using System.Windows.Media;
+using Winleafs.Models.Models;
+using Winleafs.Models.Models.Layouts;
+using Winleafs.Wpf.Api;
+using Winleafs.Wpf.Api.Effects;
+using Winleafs.Wpf.Api.Layouts;
+using Winleafs.Wpf.Helpers;
+using Winleafs.Wpf.Views.Popup;
+
+namespace Winleafs.Wpf.Views.Layout
+{
+
+	/// <summary>
+	/// Interaction logic for CreateEffectWindow.xaml
+	/// </summary>
+	public partial class CreateEffectWindow : Window
+    {
+
+		private IList<FrameListItem> _frameListItems;
+		private CustomEffect _customEffect;
+		private Frame _currentFrame;
+		private Dictionary<uint, SolidColorBrush> _rgbToBrushMap;
+		private static readonly Regex _numericRegex = new Regex("[^0-9.]");
+		private static readonly string _frameNameFormat = Layout.Resources.Frame + " {0}";
+
+		public CreateEffectWindow()
+        {
+            InitializeComponent();
+
+			LayoutDisplay.EnableClick();
+			LayoutDisplay.InitializeResizeTimer();
+			LayoutDisplay.DrawLayout();
+            LayoutDisplay.DisableColorTimer();
+			LayoutDisplay.MultiSelectEnabled = false;
+			LayoutDisplay.PanelClicked += LayoutDisplay_PanelClicked;
+			ColorPicker.SelectedColor = Colors.White;
+
+            if (UserSettings.Settings.ActiveDevice.CustomEffect != null)
+            {
+                var serialized = JsonConvert.SerializeObject(UserSettings.Settings.ActiveDevice.CustomEffect); //Deep copy the custom effect when editing
+                _customEffect = JsonConvert.DeserializeObject<CustomEffect>(serialized);
+			}
+            else
+            {
+                _customEffect = new CustomEffect();
+			}
+
+			BuildRgbToBrushMap();
+
+			BuildFrameList();
+		}
+				
+		private void LayoutDisplay_PanelClicked(object sender, System.EventArgs e)
+		{
+			var drawablePanel = sender as DrawablePanel;
+
+			if (drawablePanel != null)
+			{
+				var color = ColorPicker.SelectedColor ?? Colors.Black;
+
+				//Try to find a brush for the selected color
+				var rgb = ColorFormatConverter.ToRgb(color);
+
+				if (!_rgbToBrushMap.TryGetValue(rgb, out var brush))
+				{
+					AddToRgbToBrushMap(color);
+				}
+
+				//Color the panel and update the color for the panel on the Frame
+				drawablePanel.Polygon.Fill = _rgbToBrushMap[rgb];
+				_currentFrame.PanelColors[drawablePanel.PanelId] = rgb;
+			}
+		}
+
+		private void AddFrame_Click(object sender, RoutedEventArgs e)
+		{
+			AddNewFrame();
+		}
+
+		private void RemoveFrame_Click(object sender, RoutedEventArgs e)
+		{
+			//Always leave 1 frame
+			if (FrameListBox.Items.Count < 2)
+			{
+				return;
+			}
+
+			var lastFrameListItem = _frameListItems.Last(); 
```

Comment:
Need to check this within a functional review but wouldn't it be nicer to deleted the frame that is currently selected instead of always the last frame?

Reply:
I n the end I decided that the need to remove a frame from the middle was unlikely and that the likely use case would be to remove the last one if you added too many. Let me know after you have had a chance to play with it.